### PR TITLE
Example

### PR DESCRIPTION
**Pull request by sheenobu**
This is both the implementatino of the core structures and an example of a 'contains' plugin based on some example yaml.

```
m.ServeIRC(ctx, testMessage("hello world"), &w)
m.ServeIRC(ctx, testMessage("world !Kris"), &w)
m.ServeIRC(ctx, testMessage("world !@Kris"), &w)
```

### CODE REVIEW

File: pkg/tt/handler.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+type Handler interface {
+	ServeIRC(ctx context.Context, msg *Message, w MessageWriter)
+}
+
+// HandlerFunc is the functional alias for the Handler interface
+type HandlerFunc func(ctx context.Context, msg *Message, w MessageWriter)
+
+// ServeIRC handles incoming IRC messages
+func (h HandlerFunc) ServeIRC(ctx context.Context, msg *Message, w MessageWriter) {
+	h(ctx, msg, w)
```
Comment (krisnova) [inline pkg/tt/handler.go:15]: Can we use a more verbose function name please?

File: pkg/tt/mux/mux.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+type mux struct {
+	l       sync.RWMutex
+	rules   []tt.Rule
+	plugins map[int64]tt.Plugin
+}
+
+func (mx *mux) Register(p tt.Plugin) {
+	rl := p.Rules()
+
+	mx.l.Lock()
```
Comment (krisnova) [inline pkg/tt/mux/mux.go:33]: This looks like a mutex - is this a mutex?

Comment (sheenobu) [reply pkg/tt/mux/mux.go:33]: Yes. RWMutex. Full Read-Write lock during registration of plugins (usually on system startup) and Read lock during message routing. Shouldn't be an issue.

File: pkg/tt/handler.go (hunk 1 above)
Comment (sheenobu) [reply pkg/tt/handler.go:15]: you are talking about h? it's just the 'self' or 'this' for the HandlerFunc type alias.

Comment (krisnova) [reply pkg/tt/handler.go:15]: oh I didn't even see that - thank you!

**Comment by krisnova** [comment]
@sheenobu LGTM - Please remove the `don't merge` title whenever you are done

**Pull request merged**